### PR TITLE
[42] [43] As a user, I can refresh token when it expired

### PR DESCRIPTION
**Pull request by manh-t**
https://github.com/manh-t/flutter-survey/issues/42
https://github.com/manh-t/flutter-survey/issues/43

## What happened 👀

Backend + Integrate

## Insight 📝

Refactor naming `OauthTokenRequest` -> `OauthLoginRequest` to consistent with context

## Proof Of Work 📹

I print logs during calling apis to easily track progress. As you can see, we only call refreshing token 1 time(`REFRESH_TOKEN REFRESHING TOKENNNNNNNNNNNNNN`) although we call 2 apis at the same time, and 2 apis are also failed🤓 

```
I/flutter (28996): REFRESH_TOKEN old token on request: Bearer v_x7Z6Rtzte_iZbQ861o8w5E1O5r4srqUXDN5jXCpAY
I/flutter (28996): REFRESH_TOKEN refresh token on request: ElVk79zlRBbNE1RO1k-EBr6RazKPto1WmQx2LJR116g
I/flutter (28996): REFRESH_TOKEN old token on request: Bearer v_x7Z6Rtzte_iZbQ861o8w5E1O5r4srqUXDN5jXCpAY
I/flutter (28996): REFRESH_TOKEN refresh token on request: ElVk79zlRBbNE1RO1k-EBr6RazKPto1WmQx2LJR116g
I/flutter (28996): *** Request ***
I/flutter (28996): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/me
I/flutter (28996): method: GET
I/flutter (28996): responseType: ResponseType.json
I/flutter (28996): followRedirects: true
I/flutter (28996): connectTimeout: 30000
I/flutter (28996): sendTimeout: 0
I/flutter (28996): receiveTimeout: 30000
I/flutter (28996): receiveDataWhenStatusError: true
I/flutter (28996): extra: {}
I/flutter (28996): headers:
I/flutter (28996):  Authorization: Bearer v_x7Z6Rtzte_iZbQ861o8w5E1O5r4srqUXDN5jXCpAY
I/flutter (28996): data:
I/flutter (28996): {}
I/flutter (28996):
I/flutter (28996): *** Request ***
I/flutter (28996): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=10
I/flutter (28996): method: GET
I/flutter (28996): responseType: ResponseType.json
I/flutter (28996): followRedirects: true
I/flutter (28996): connectTimeout: 30000
I/flutter (28996): sendTimeout: 0
I/flutter (28996): receiveTimeout: 30000
I/flutter (28996): receiveDataWhenStatusError: true
I/flutter (28996): extra: {}
I/flutter (28996): headers:
I/flutter (28996):  Authorization: Bearer v_x7Z6Rtzte_iZbQ861o8w5E1O5r4srqUXDN5jXCpAY
I/flutter (28996): data:
I/flutter (28996): {}
I/flutter (28996):
I/flutter (28996): REFRESH_TOKEN error code: 401
I/flutter (28996): REFRESH_TOKEN old token: Bearer v_x7Z6Rtzte_iZbQ861o8w5E1O5r4srqUXDN5jXCpAY
I/flutter (28996): REFRESH_TOKEN REFRESHING TOKENNNNNNNNNNNNNN
I/flutter (28996): REFRESH_TOKEN error code: 401
I/flutter (28996): REFRESH_TOKEN old token: Bearer v_x7Z6Rtzte_iZbQ861o8w5E1O5r4srqUXDN5jXCpAY
2
I/flutter (28996): REFRESH_TOKEN on error: Bearer aEczkj07sS0v2DJ7gQzs2x0q1ISbMiVCaSnVaxpWThI
I/flutter (28996): REFRESH_TOKEN old token on request: Bearer aEczkj07sS0v2DJ7gQzs2x0q1ISbMiVCaSnVaxpWThI
I/flutter (28996): REFRESH_TOKEN refresh token on request: dlkdAHJR4n7waOXGhGmfmEzhgL6vRJ2GFwuIpFOmBKg
I/flutter (28996): REFRESH_TOKEN old token on request: Bearer aEczkj07sS0v2DJ7gQzs2x0q1ISbMiVCaSnVaxpWThI
I/flutter (28996): REFRESH_TOKEN refresh token on request: dlkdAHJR4n7waOXGhGmfmEzhgL6vRJ2GFwuIpFOmBKg
I/flutter (28996): *** Request ***
I/flutter (28996): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=10
I/flutter (28996): method: GET
I/flutter (28996): responseType: ResponseType.json
I/flutter (28996): followRedirects: true
I/flutter (28996): connectTimeout: 30000
I/flutter (28996): sendTimeout: 0
I/flutter (28996): receiveTimeout: 30000
I/flutter (28996): receiveDataWhenStatusError: true
I/flutter (28996): extra: {}
I/flutter (28996): headers:
I/flutter (28996):  Authorization: Bearer aEczkj07sS0v2DJ7gQzs2x0q1ISbMiVCaSnVaxpWThI
I/flutter (28996): data:
I/flutter (28996): {}
I/flutter (28996):
I/flutter (28996):
I/flutter (28996): *** Request ***
I/flutter (28996): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/me
I/flutter (28996): method: GET
I/flutter (28996): responseType: ResponseType.json
I/flutter (28996): followRedirects: true
I/flutter (28996): connectTimeout: 30000
I/flutter (28996): sendTimeout: 0
I/flutter (28996): receiveTimeout: 30000
I/flutter (28996): receiveDataWhenStatusError: true
I/flutter (28996): extra: {}
I/flutter (28996): headers:
I/flutter (28996):  Authorization: Bearer aEczkj07sS0v2DJ7gQzs2x0q1ISbMiVCaSnVaxpWThI
I/flutter (28996): data:
I/flutter (28996): {}
I/flutter (28996):
I/flutter (28996): *** Response ***
I/flutter (28996): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=10
I/flutter (28996): statusCode: 200
I/flutter (28996): headers:
I/flutter (28996):  connection: keep-alive
I/flutter (28996):  cache-control: max-age=0, private, must-revalidate
I/flutter (28996):  transfer-encoding: chunked
I/flutter (28996):  date: Tue, 08 Jun 2021 13:38:56 GMT
I/flutter (28996):  vary: Accept-Encoding, Origin
I/flutter (28996):  content-encoding: gzip
I/flutter (28996):  strict-transport-security: max-age=31536000; includeSubDomains
I/flutter (28996):  referrer-policy: strict-origin-when-cross-origin
I/flutter (28996):  x-permitted-cross-domain-policies: none
I/flutter (28996):  content-type: application/json; charset=utf-8
I/flutter (28996):  x-xss-protection: 1; mode=block
I/flutter (28996):  server: Cowboy
I/flutter (28996):  x-request-id: d072d558-b95d-4898-83cf-f722f0a042a4
I/flutter (28996):  x-download-options: noopen
I/flutter (28996):  x-runtime: 0.019809
I/flutter (28996):  etag: W/"55c118b30707aee8f6846810edbf16ff"
I/flutter (28996):  via: 1.1 vegur
I/flutter (28996):  x-frame-options: SAMEORIGIN
I/flutter (28996):  x-content-type-options: nosniff
I/flutter (28996): Response Text:
I/flutter (28996): {"data":[{"id":"d5de6a8f8f5f1cfe51bc","type":"survey","attributes":{"title":"Scarlett Bangkok","description":"We'd love ot hear from you!","thank_email_above_threshold":"<span style=\"font-family:arial,helvetica,sans-serif\"><span style=\"font-size:14px\">Dear {name},<br /><br />Thank you for visiting Scarlett Wine Bar &amp; Restaurant at Pullman Bangkok Hotel G &nbsp;and for taking the time to complete our guest feedback survey!<br /><br />Your feedback is very important to us and each survey is read individually by the management and owners shortly after it is sent. We discuss comments and suggestions at our daily meetings and use them to constantly improve our services.<br /><br />We would very much appreciate it if you could take a few more moments and review us on TripAdvisor regarding your recent visit. By <a href=\"https://www.tripadvisor.com/Restaurant_Review-g293916-d2629404-Reviews-Scarlett_Wine_Bar_Restaurant-Bangkok.html\">clicking here</a> you will be directed to our page.&nbsp;<br /><br />Thank
I/flutter (28996):
I/flutter (28996): *** Response ***
I/flutter (28996): uri: https://nimble-survey-web-staging.herokuapp.com/api/v1/me
I/flutter (28996): statusCode: 200
I/flutter (28996): headers:
I/flutter (28996):  connection: keep-alive
I/flutter (28996):  cache-control: max-age=0, private, must-revalidate
I/flutter (28996):  transfer-encoding: chunked
I/flutter (28996):  date: Tue, 08 Jun 2021 13:38:58 GMT
I/flutter (28996):  vary: Accept-Encoding, Origin
I/flutter (28996):  content-encoding: gzip
I/flutter (28996):  strict-transport-security: max-age=31536000; includeSubDomains
I/flutter (28996):  referrer-policy: strict-origin-when-cross-origin
I/flutter (28996):  x-permitted-cross-domain-policies: none
I/flutter (28996):  content-type: application/json; charset=utf-8
I/flutter (28996):  x-xss-protection: 1; mode=block
I/flutter (28996):  server: Cowboy
I/flutter (28996):  x-request-id: 8417d4be-0d10-4643-b565-2e48d05c8782
I/flutter (28996):  x-download-options: noopen
I/flutter (28996):  x-runtime: 0.013482
I/flutter (28996):  etag: W/"9fdff0b5bb6e3337519ed5feae6ec4bd"
I/flutter (28996):  via: 1.1 vegur
I/flutter (28996):  x-frame-options: SAMEORIGIN
I/flutter (28996):  x-content-type-options: nosniff
I/flutter (28996): Response Text:
I/flutter (28996): {"data":{"id":"2","type":"user","attributes":{"email":"dev@nimblehq.co","avatar_url":"https://api.adorable.io/avatar/dev@nimblehq.co"}}}
```

### CODE REVIEW

**Review by markgravity — APPROVED** [review]
LGTM 🤖

File: lib/api/service/oauth_service.dart [hunk 1]
```diff
@@ -21,11 +22,15 @@ abstract class OauthService {
 
   @POST(OauthApi.oauthToken)
   Future<BaseHttpResponse<OauthTokenResponse>> login(
-      @Body() OauthTokenRequest body);
+      @Body() OauthLoginRequest body);
```
Comment (sleepylee) [inline lib/api/service/oauth_service.dart:25]: in general of this PR: it's `OAuth` fyi, no `Oauth` 😛  
https://en.wikipedia.org/wiki/OAuth

File: lib/di/interceptor/app_interceptor.dart [hunk 2]
```diff
@@ -1,19 +1,95 @@
...
       options.headers.putIfAbsent(
-          'Authorization', () => _sharedPreferenceHelper.authToken);
+          HEADER_AUTHORIZATION, () => "${sharedPreferenceHelper.authToken}");
     }
     return super.onRequest(options, handler);
   }
+
+  @override
+  void onError(DioError err, ErrorInterceptorHandler handler) async {
+    if (err.response?.statusCode == 403 || err.response?.statusCode == 401) {
```
Comment (sleepylee) [inline lib/di/interceptor/app_interceptor.dart:40]: 403 is unauthroized, it doesn't seem like related to the cause of our token expired with our current survey backend? 🤔

**Review by sleepylee — APPROVED** [review]

**Comment by sleepylee** [comment]
just a few minor things, renaming them and it should be good to go

File: lib/api/service/oauth_service.dart (hunk 1 above)
Comment (manh-t) [reply lib/api/service/oauth_service.dart:25]: 😱

File: lib/di/interceptor/app_interceptor.dart (hunk 2 above)
Comment (manh-t) [reply lib/di/interceptor/app_interceptor.dart:40]: oh, sorry I put it to test but forgot removing 😓 Will remove now

**Pull request merged**